src/thread_pool.cpp

```diff
@@ -75,33 +77,35 @@ auto thread_pool::shutdown() noexcept -> void
     }
 }
 
-auto thread_pool::executor(std::stop_token stop_token, std::size_t idx) -> void
+auto thread_pool::executor(std::size_t idx) -> void
 {
     if (m_opts.on_thread_start_functor != nullptr)
     {
         m_opts.on_thread_start_functor(idx);
     }
 
-    while (!stop_token.stop_requested())
+    // Process until shutdown is requested and the total number of tasks reaches zero.
+    while (!m_shutdown_requested.load(std::memory_order::acquire) || m_size.load(std::memory_order::acquire) > 0)
     {
-        // Wait until the queue has operations to execute or shutdown has been requested.
-        while (true)
+        std::unique_lock<std::mutex> lk{m_wait_mutex};
+        m_wait_cv.wait(
+            lk,
+            [&] {
+                return m_size.load(std::memory_order::acquire) > 0 ||
+                       m_shutdown_requested.load(std::memory_order::acquire);
+            });
+
+        // Process this batch until the queue is empty.
+        while (!m_queue.empty())
```

Comment:
I've switched the logic here to drain the queue without the condition variable being invoked each time between executions of tasks